syntax = "proto2";

//////////////////////////////////////////////////////////////////////
// IMPORTANT:                                                       //
// When changing this file, you MUST re-run the following:          //
// protoc --python_out=scripts api/clover.proto --proto_path=api    //
//////////////////////////////////////////////////////////////////////

// Command stream.
// Delivered over a bytestream. Each incoming request is processed in sequence and will have its responses delivered in
// the same sequence.

// Every request sent to the server is an instance of this parent Request. The server then examines the payload to
// determine the exact command specified.
message Request {
  oneof payload {
    SubscribeDataStreamRequest subscribe_data_stream = 1;
    IdentifyClientRequest identify_client = 6;
    ResetValvePositionRequest reset_valve_position = 2;
    LoadMotorSequenceRequest load_motor_sequence = 3;
    StartSequenceRequest start_sequence = 4;
    HaltSequenceRequest halt_sequence = 5;
    StartThrottleClosedLoopRequest start_throttle_closed_loop = 7; // NEW: Renamed and handles trace
    SetControllerStateRequest set_controller_state = 8;
  }
}

// Successful if err is missing.
message Response {
  optional string err = 1;
}

// Subscribe client IP to data stream
message SubscribeDataStreamRequest {}

// Names client
message IdentifyClientRequest {
  required ClientType client = 1;
}

enum ClientType {
  UNKNOWN_CLIENT = 1;
  GNC = 2;
  DAQ = 3;
}

// Throttle valve control
enum Valve {
  UNKNOWN_VALVE = 0;
  FUEL = 1;
  LOX = 2;
}

// ResetValvePosition
message ResetValvePositionRequest {
  required Valve valve = 1;
  required float new_pos_deg = 2;
}

// Control sequences
message ControlTrace {
  required uint32 total_time_ms = 1;
  repeated Segment segments = 2;
}

message Segment {
  required uint32 start_ms = 1;
  required uint32 length_ms = 2;
  oneof type {
    LinearSegment linear = 3;
    SineSegment sine = 4;
  }
}

message LinearSegment {
  required float start_val = 1;
  required float end_val = 2;
}

message SineSegment {
  required float offset = 1;
  required float amplitude = 2;
  required float period = 3;
  required float phase_deg = 4;
}

message LoadMotorSequenceRequest {
  optional ControlTrace fuel_trace = 1;
  optional ControlTrace lox_trace = 2;
}

// StartSequence
message StartSequenceRequest {}

// HaltSequence
message HaltSequenceRequest {}


// NEW: StartThrottleClosedLoop
message StartThrottleClosedLoopRequest {
  optional ControlTrace thrust_trace = 1;
}

// Change Controller State
message SetControllerStateRequest {
  required SystemState state = 1;
}

// Data stream.
// Delivered in individual packets on a regular interval, only from clover to external listeners.

// Controller State definition
enum SystemState {
  STATE_IDLE = 0;
  STATE_SEQUENCE = 1;
  STATE_CLOSED_LOOP_THROTTLE = 2;
  STATE_ABORT = 3;
  STATE_CALIBRATION = 4;
}

message DataPacket {
  required float time = 1;
  required uint32 data_queue_size = 2;

  required ValveStatus fuel_valve = 3;
  required ValveStatus lox_valve = 4;

  required Sensors sensors = 5;

  required SystemState state = 6;
  required bool is_abort = 7;
  required uint32 sequence_number = 8;

  optional CalibrationData calibration_data = 9;


}

message ValveStatus {
  required bool enabled = 1;
  required float target_pos_deg = 2;
  required float driver_setpoint_pos_deg = 3;
  required float encoder_pos_deg = 4;
}

message Sensors {
  optional float pt102 = 1;
  optional float pt103 = 2;
  optional float pt202 = 3;
  optional float pt203 = 4;
  optional float ptf401 = 5;
  optional float pto401 = 6;
  optional float ptc401 = 7;
  optional float ptc402 = 8;
}


message CalibrationData {
  required float fuel_pos = 1;
  required float fuel_pos_enc = 2;
  required bool fuel_found_hardstop = 3;
  required float fuel_hardstop_pos = 4;
  required float lox_pos = 5;
  required float lox_pos_enc = 6;
  required bool lox_found_hardstop = 7;
  required float lox_hardstop_pos = 8;

  required int32 cal_phase = 9;
  required int32 rep_count = 10;

  required float fuel_err = 11;
  required float lox_err = 12;

  required float fuel_target_position = 13;
  required float lox_target_position = 14;

}
